diff --git a/drivers/net/wireless/ath/ath11k/debugfs.c b/drivers/net/wireless/ath/ath11k/debugfs.c
index 436197113ced..ec4683695e73 100644
--- a/drivers/net/wireless/ath/ath11k/debugfs.c
+++ b/drivers/net/wireless/ath/ath11k/debugfs.c
@@ -132,7 +132,9 @@ void ath11k_debugfs_fw_stats_process(struct ath11k *ar, struct ath11k_fw_stats *
 				total_vdevs_started += ar->num_started_vdevs;
 		}
 
-		is_end = ((++num_vdev) == total_vdevs_started);
+		is_end = ((++num_vdev) >= total_vdevs_started);
+		if (num_vdev > total_vdevs_started)
+			ath11k_warn(ab, "spurious stats");
 
 		list_splice_tail_init(&stats->vdevs,
 				      &ar->fw_stats.vdevs);
